src/chain/shared_state.rs

```diff
@@ -627,44 +626,58 @@ impl SectionProofChain {
         }
     }
 
-    fn validate_next_block(&self, block: &SectionProofBlock) -> bool {
-        let last = self.last_public_key_info();
-        if block.version() != last.version() + 1 {
-            log_or_panic!(
-                LogLevel::Error,
-                "Invalid next block version - not successor (last: {}, next: {})",
-                last.version(),
-                block.version()
-            );
-            return false;
+    fn validate_next_block(
+        &self,
+        last: &SectionKeyInfo,
+        next: &SectionProofBlock,
+    ) -> Result<(), NextBlockError> {
+        if next.version() != last.version() + 1 {
+            return Err(NextBlockError::Version {
+                actual: next.version(),
+                expected: last.version() + 1,
+            });
         }
 
-        if !block.prefix().is_compatible(last.prefix()) {
-            log_or_panic!(
-                LogLevel::Error,
-                "Invalid next block prefix - not compatible (last: {:?}, next: {:?})",
-                last.prefix(),
-                block.prefix()
-            );
-            return false;
+        if !next.prefix().is_compatible(last.prefix())
+            || next.prefix().bit_count() > last.prefix().bit_count() + 1
+        {
+            return Err(NextBlockError::Prefix {
+                actual: *next.prefix(),
+                expected: *last.prefix(),
+            });
         }
 
-        if block.prefix() != last.prefix()
-            && block.prefix().bit_count() != last.prefix().bit_count() + 1
-        {
-            log_or_panic!(
-                LogLevel::Error,
-                "Invalid next block prefix - differs by more than one bit (last: {:?}, next: {:?})",
-                last.prefix(),
-                block.prefix()
-            );
-            return false;
+        if !next.verify_with_pk(*last.key()) {
+            return Err(NextBlockError::Signature);
         }
 
-        true
+        Ok(())
     }
 }
 
+#[derive(Error, Debug)]
+enum NextBlockError {
```

Comment:
Seem way too complicated for what we need, after all we essentially don't expect to see that message.

```
fn validate_next_block(...) -> bool;

     pub fn push(&mut self, block: SectionProofBlock) {
        if self.validate_next_block(self.last_public_key_info(), &block) {
            self.blocks.push(block);
        } else {
            log_or_panic!(LogLevel::Error, "Invalid next block {:?} -> {:?}", self.last_public_key_info(), block),
        }
    }

Reply:
Fair enough. Will simplify it.